src/shims/unix/linux/epoll.rs

```diff
@@ -507,10 +594,31 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                         let event_instance =
                             EpollEventInstance::new(flags, epoll_event_interest.data);
                         ready_list.insert(epoll_key, event_instance);
+                        // Edge-triggered notification only notify one thread even if there are
+                        // multiple threads block on the same epfd.
+                        let epfd = this.machine.fds.get(epoll_event_interest.epfd).unwrap();
+                        // FIXME: We can randomly pick a thread to unblock.
+
+                        // This unwrap can never fail because if the current epoll instance were
+                        // closed and its epfd value reused, the upgrade of weak_epoll_interest
+                        // above would fail. This guarantee holds because only the epoll instance
+                        // holds a strong ref to epoll_interest.
+                        if let Some(thread_id) =
+                            epfd.downcast::<Epoll>().unwrap().thread_id.borrow_mut().pop()
+                        {
+                            waiter.push(thread_id);
+                        };
                     }
                 }
             }
         }
+        waiter.sort();
+        waiter.dedup();
+        for thread_id in waiter {
+            if this.has_blocked_on_epoll(thread_id) {
+                this.unblock_thread(thread_id, BlockReason::Epoll)?;
+            }
```

Comment:
It turned out that it might be possible for a thread to wake up before we call unblock, so we'd need an additional check here.

Reply:
That sounds like the blocked thread bookkeeping is going wrong somewhere. Who else is waking up this thread?

Reply:
Ah right, there is a problem with the thread bookkeeping. The thread is woken up because of timeout, I should clean it up when timeout has reached.